### PR TITLE
Add explicit string casts for common/easy cases

### PR DESCRIPTION
**Pull request by bradcray**
When we do a cast-to-string (say, in an error message) and the type
doesn't support such a cast, Chapel calls into the general-purpose
stringify() routine which is very cool in that it converts any type to
a string, but is very expensive in that it does it by virtue of
writing the value to string memory using some pretty nontrivial I/O code.

This PR takes a few common types that are compact and easy to write to
strings (ranges, rectangular domains, byteIndices, codepointIndices)
and injects string casts for them to avoid these code paths.  For domains
specifically, this adds a `.doiToString()` routine to the optional interface.
The net effect of these changes reverses the increase in generated code
that stemmed from #17628 and then some, resulting in a ~4000 line savings
in generated code for the stream benchmark with checks on.

While adding tests for these string casts, I found and fixed a few errors
in converting ranges to strings:

* for single-element enum ranges, we weren't generating the same `lo..<lo`
  pattern that we do for I/O
* similarly, for `..hi` ranges, we were tacking on an `align 0` even though
  we don't normally print alignment for non-strided ranges

These oversights suggest that it'd be nice to better unify default I/O and
cast-to-string code to avoid repeated code and cases where their behavior
diverges like this.

I believe this will resolve https://github.com/Cray/chapel-private/issues/2074
but am not marking it as such here until we verify in nightly testing.

### CODE REVIEW

**Comment by bradcray** [comment]
@ronawho: Are you interested in reviewing this, since you put me on the task?

File: modules/internal/ChapelArray.chpl [hunk 1]
```diff
@@ -2346,6 +2346,15 @@ module ChapelArray {
       return _value.dsiIteratorYieldsLocalElements();
     }
 
+    operator :(val: _domain, type t: string) {
+      use Reflection;
+      if canResolveMethod(val._value, "dsiToString") {
+        return val._value.dsiToString();
```
Comment (ronawho) [inline modules/internal/ChapelArray.chpl:2352]: Should `dsiToString` be added to the dsi tech note?

**Review by ronawho — APPROVED** [review]

File: modules/internal/ChapelArray.chpl (hunk 1 above)
Comment (bradcray) [reply modules/internal/ChapelArray.chpl:2352]: Hmm, your question is making me realize that this should really be a `.doi...` routine given that it's optional (and may not make sense for something like a sparse or associative domain where the generated code size of converting to a string is likely to dwarf the general `stringify()` routine.  And then the question still holds, except that it seems we don't document many of the doi routines in that file...?  [mostly, I feel like that file is waiting for an interface to replace it]

Comment (bradcray) [reply modules/internal/ChapelArray.chpl]: (I did end up adding it, though from an account where I didn't have DCO signed up, so need to clean up.  While adding more testing, I also found some pre-existing bugs in the range cast code which were easy to fix, but will require a re-test.

**Comment by bradcray** [comment]
@ronawho: I'm doing one more semi-paranoid check on gasnet testing, but am planning to merge tonight in order to see the impact on an otherwise quiet night of testing.  You may want to look at the commits between your last review and this one—I'd call them trivial, but still think they're worth a post-commit review.

**Pull request merged**